compiler/src/dotty/tools/dotc/core/TypeComparer.scala

```diff
@@ -403,6 +408,16 @@ class TypeComparer(initctx: Context) extends ConstraintHandling[AbsentContext] {
           val gbounds2 = gadtBounds(tp2.symbol)
           (gbounds2 != null) &&
             (isSubTypeWhenFrozen(tp1, gbounds2.lo) ||
+              (tp1 match {
+                case tp1: NamedType if ctx.gadt.contains(tp1.symbol) =>
+                  // Note: since we approximate constrained types only with their non-param bounds,
+                  // we need to manually handle the case when we're comparing two constrained types,
+                  // one of which is constrained to be a subtype of another.
+                  // We do not need similar code in fourthTry, since we only need to care about
+                  // comparing two constrained types, and that case will be handled here first.
+                  ctx.gadt.isLess(tp1.symbol, tp2.symbol) && GADTusage(tp1.symbol) && GADTusage(tp2.symbol)
```

Comment:
@smarter you might want to take a look at this and its original commit (the first in history). It seems to me that we might be missing an analogous case when comparing two TypeParamRefs in frozen constraint. Note that when _not_ in frozen constraint, `TypeComparer` will try to record a new bound, which will be a no-op.

Reply:
Reminder to self post-coversation w/ @smarter: create an issue for this after this is merged.